lib/compress.js

```diff
@@ -4499,13 +4501,38 @@ merge(Compressor.prototype, {
     });
 
     OPT(AST_Arrow, function(self, compressor){
-        if (self.body.length === 1 && self.body[0] instanceof AST_Return) {
+        if (!(self.body instanceof AST_Node)) tighten_body(self.body, compressor);
+        if (compressor.option("arrows")
+            && self.body.length == 1
+            && self.body[0] instanceof AST_Return) {
             var value = self.body[0].value;
             self.body = value ? value : [];
         }
         return self;
     });
 
+    OPT(AST_Function, function(self, compressor){
+        tighten_body(self.body, compressor);
+        if (compressor.option("arrows")
+            && compressor.option("ecma") >= 6
+            && !self.name
+            && !self.async
```

Comment:
Done in https://github.com/mishoo/UglifyJS2/pull/2154/files#diff-6235f1a4bcfe8aceaea347a4c364ecc0R238